### ISSUE TITLE
Does Paket have any culpability in VS intellisense failing issue?

### ISSUE THREAD

**Issue by jackfoxy** [fsprojects/Paket#3270]
Before I blast this message publicly, I would like some feedback from people more familiar with the intricacies of paket/dotnet. I have only every used this combination for F# projects, and never pure nuget/dotnet. Thanks.

The message:

I had high hopes for F# developer tools in VS 17.8 Preview 2, and at first I thought all was well. I was wrong. Preview 2 and now Preview 3 is an absolute shit show.

You never know if intellisense is going to work or not, but you can be certain of one thing. If it is working, it will soon break. F12 works everywhere, then doesn't work. 

I don't even know how to begin a bug report on this. If you can get intellisense and F12 to work, sooner or later you will type or paste in a code snippet that does not build and that is the end. Closing and reopening the file does not fix it. Closing and reopening VS does not fix it. Rebooting does not fix it.

The only thing that might work (and this doesn't even always work) is:

close vs
git clean -xfd
paket update -f (I think install might work better, but I have not experimented)
dotnet restore

Is the problem at the juncture of Roslyn tools and dotnet SDK?

Migrating F# development to Roslyn tools was probably a mistake in retrospect. What did F# developers get out of it? Answer: light bulbs.

Oh, and guess what? When everything else about the development tools is broken, the light bulbs don't work either.

What I'm talking about is real world solutions in production breaking the tools.

Everyone at Microsoft involved with F# needs to stop whatever they are working on and fix the developer tools. 

I will stop now before I write something I regret.

**Comment by cartermp** [comment]
@jackfoxy Paket is not supported with implicit restore at design-time, so you must restore before you open in Visual Studio. Otherwise, tooling will simply not light up because references cannot be resolved. This is due to design constraints both in the project system and in Paket, which are very difficult to reconcile on either side. This general problem isn't new to 15.8, though if you've first adopted paket for larger solutions with 15.8 previews it's understandable that it would surface these problems for you.

> Is the problem at the juncture of Roslyn tools and dotnet SDK?

Doesn't sound like it. We also dogfood our own codebase with a mix of .NET SDK and non-SDK and do not encounter what you're describing. We do not use Paket, though.

> You never know if intellisense is going to work or not, but you can be certain of one thing. If it is working, it will soon break. F12 works everywhere, then doesn't work.
>
> I don't even know how to begin a bug report on this. If you can get intellisense and F12 to work, sooner or later you will type or paste in a code snippet that does not build and that is the end. Closing and reopening the file does not fix it. Closing and reopening VS does not fix it. Rebooting does not fix it.

Unfortunately, this is not specific enough to be actionable. I've never personally encountered pasting a code snippet that fails to compile and then having all tools fail across the entire codebase (nor has this been reported before), but if you can reproduce this, please do send a bug report, as that would be a high priority issue to fix.

Same goes for Go to Definition, unless [Go to Metadata from the VFPT extension](https://github.com/Microsoft/visualfsharp/issues/1939) (Note: not officially supported) is what you are referring to. This is an open feature request with a candidate PR, but that PR has tremendous issues where the output often fails to write a metadata file with complete information. So it's not merged since it's quite broken w.r.t the actual result it produces.

Really sorry that you're seeing these issues, but unfortunately, without reproducible scenarios for what you've described, no action can be taken unless it can be observed in our daily work. And with regards to Paket integration, serious work in the project system and/or Paket will be required before any acceptable experience surrounding implicit restore can occur.

**Comment by forki** [comment]
I don't know how much of this is due to paket - I can only try to analyze concrete repros. ;-(

**Comment by forki** [comment]
> Paket is not supported with implicit restore at design-time

That's not true. That actually worked long time ago. Maybe we regressed.

Phillip Carter <notifications@github.com> schrieb am Mi., 27. Juni 2018,
20:20:

**Comment by cartermp** [comment]
Interesting. I wonder if this was back when the SDK project system was still super new, and it happened to work before the proper machinery was in place?

**Comment by 0x53A** [comment]
It would be kind of awesome if you could detect that scenario and pop open a yellow information bar at the top.

This is how rider does it since end of 2017: https://twitter.com/0x53A/status/924421382653476864?s=09

**Comment by forki** [comment]
No. We just did what @davkean told us at that time and it worked.

Lukas Rieger <notifications@github.com> schrieb am Mi., 27. Juni 2018,
21:34:

**Comment by 0x53A** [comment]
That was targeted at @cartermp and @davkean, paket can't and shouldn't open any message boxes.

**Comment by forki** [comment]
Seriously: we were told that referencing a targets file and loading the PackageReferences in memory is a good option and would work with VS. We did that and it worked.

Don't know why it doesn't anymore. Did not test for long time.

**Comment by forki** [comment]
@davkean I remember there was a change in the order of the MSBuild targets for dependencies restore. Maybe we need to hook it up in a new one for background restore to see it?

**Comment by davkean** [comment]
@forki I've indicated many times that implicit restore with paket requires paket to do more work than they are currently, my last time was here: https://twitter.com/davkean/status/1003972527926886401. You've acknowledged that in the past, but keep coming back to this "VS has changed behavior part".

@0x53A I really like how Rider has done it, we should be better about this. At minimum paket should probably put an error in the error list to manually run it.

**Comment by forki** [comment]
@davkean I did not say VS has changed/regressed. But I think the combination of paket/NuGet targets/VS did. In our very old email thread we already had it working. I even remember tweeting gifs about it. Imho a plugin in VS is just not a good solution. We were always looking for a MSBuild-only solution, so that things work ootb for everyone.

**Comment by forki** [comment]
@davkean it seems there is a bit of a misunderstanding here. I was referring to @cartermp. He said you need to restore before opening VS. But that should work since it's "initial restore" and does not need file watchers.
In your tweet I think you were referring to the situation when paket files were touched and this does not trigger an additional background restore. For that part we set a MSBuild property RestoreSuccess to false. That seemed to trigger it, but maybe it's not enough. I was wondering if nuking the assets file would work. Do you know if that's watched? Thanks

**Comment by davkean** [comment]
Are we talking about the same scenario?

1. Open project in VS
2. Edit paket.dependencies/paket.references and save.
3. ???

How do you delete the assets file in step 3? What code of yours is running in this situation?

**Comment by forki** [comment]
I see 2 scenarios that I hope should work ootb without a plugin.

1) 

* git clone
* open VS
* restore is triggered and intellisense works 

2) 
* after 1) edit paket.dependencies + paket.references
* call paket.exe install from cmd line outside of vs
* (paket nukes assets file) 
* VS triggers background restore and intellisense works 


I think if we get these two working reliably then it would be great.

**Comment by davkean** [comment]
Two things:

In the case of 1) - does paket _itself_ (not NuGet) need to hit the network? 

In the case of 2), deleting the assets file will trigger restore - but it's heavy hammer and will result in us throwing away the world instead of just the diff. I would prefer a different mechanism, such as touching a file[1] that we're watching which will cause us to reevaluate -> design-time build -> restore if there's a change in package references.

[1] We can get to what that looks like later.

**Comment by forki** [comment]
Yes paket needs to download. Same as nuget would do. It would also need to download from other sources, since we support more than just nupkgs. 

What if we touch the csproj with a zero diff (changing the timestamp)

**Comment by forki** [comment]
Well, it has to download eventually. That's just the way it is. If someone
was calling restore before then thongs are already cached, but after git
clone it would take a while.


David Kean <notifications@github.com> schrieb am Do., 28. Juni 2018, 08:23:

**Comment by davkean** [comment]
I'm still hazy on how this works - are you saying that you are not letting NuGet download the packages? Why is that?

This is my concern; you will be running 3rd party code hitting the network in a path that is supposed to take under 200ms. Then devs start blaming VS because it's not working as expected, like the start of this thread.

**Comment by forki** [comment]
For nupkgs we could probably let nuget do parts of it (if we find a way to specify nuget source url per package). But what would that improve? Nuget can't download large packages in 200ms.

**Comment by davkean** [comment]
NuGet doesn't download packages during design-time builds. It's all background and we don't block any parts of VS based on restore.

**Comment by davkean** [comment]
As a reminder a design-time build is described here: https://github.com/dotnet/project-system/blob/master/docs/design-time-builds.md.

**Comment by davkean** [comment]
I think the best next step here is make sure we're all the same page around how things work in NuGet today, and _why_ they work as such. Based on all the perf work we've been doing, I'm extremely sensitive to 1st and 3rd party (increasing design-time build times. These affects almost all VS operations and makes the entire VS seem slow and sluggish. I have a mental model in my head of how paket _should_ plug in, but it doesn't seem to match your mental model of it. I think next steps should be set up a call and just discuss scenarios and so forth.

How's that sound?

**Comment by forki** [comment]
OK. We could probably benefit from that when PackageReferences has an attribute for the nuget source (it's important for us) and if background build sets a MSBuild property (I guess you already do that). Then I could let paket ignore the downloads and let nuget do it. 

In other situations we can't let nuget do it, since we do a lot of stuff on top - like generating load scripts for F#/C# interactive.

**Comment by vivainio** [comment]
Hey, I'm a rando Paket user that jumped on this thread.

Before expecting  anything to compile, we run ".paket/paket.exe restore" from command line in the root of our deep, multi-solution application.

There is no scenario where Visual Studio would be downloading anything. If the scenario where you edit paket.dependencies and paket.references while a solution is open, and re-do "paket restore" from command line works (i.e. VS prompts for reload), that's already good enough.

**Comment by Krzysztof-Cieslak** [comment]
Hey, I'm a rando Paket user that jumped on this thread.

Background restore works just fine in Code.

**Comment by forki** [comment]
@cartermp I can now confirm. scenario 1) from above where you just open VS after clone doesn't work anymore. It doesn't trigger paket restore anymore

**Comment by forki** [comment]
actually scratch that. I accidentally removed the gitignore paket's targets file. Scenario 1) still works and it doesn't block any UI. it works as expected. After a while the packages are restore and squigglies go away. so that's good

**Comment by forki** [comment]
@Krzysztof-Cieslak regarding VS Code. Unfortunately it only works for F# there. Omnisharp doesn't automatically restore at all, they only pop up one window per project that asks if they should restore. that's a bit annoying if you have more than zero C# projects.

**Comment by forki** [comment]
and tbf it only works because there is a plugin - and installing plugins is much more common in VS code world than in VS world.

**Comment by forki** [comment]
@davkean @enricosada quick question: does VS set a MSBuild variable when background build/restore is running? which one would that be?

**Comment by 0x53A** [comment]
> Use the DesignTimeBuild (CPS-based projects) and BuildingProject (legacy project system) properties to differentiate between when a target is run in a design-time build versus a normal build.

From https://github.com/dotnet/project-system/blob/master/docs/design-time-builds.md

**Comment by forki** [comment]
Ok, I think I found a way to allow 2):

We already create a *.csproj.paket.props file in /obj in order to put CliReferences there. I can dump the PackageReferences there as well. VS is watching the file and triggers restore/resolution. 
I didn't do that before since there is a timing issue when you run dotnet build from cmd line. In that case MsBuild doesn't see the packagesreferences since loading of props files is already done. But with yet another Msbuild switch we can keep current behaviour (PackageReferences in memory) for cmd line builds and move the references to props file for background builds. Does this sound reasonable?

**Comment by jackfoxy** [comment]
I just spent an entire day in VS 15.8 Preview3 on the big solution for my day job with no problems.


Then I switched to https://github.com/fsprojects/FSharpx.Collections/commit/37d582a5d49edbe5ce41ce23b767a2a72961fc73
and in a file of FSharp.Collections.Experimental.Tests found the project in a state where the in-solution referenced projects were not accessible by intellisense. Then I did the following:

1) exited VS
2) command line paket restore

3) opened VS, in file in question waited until intellisense had finished doing its thing

3a) Damn! still not recognizing projects

4) closed VS again, with intention to do git clean, etc.

4a) remembered "I should really try building in debug first"


5) returned to VS, waited for intellisense to complete again

5a) this time projects are recognized and intellisense is working


cc @forki  @cartermp 
I don't know if this provides you any useful information, but that's my story.

**Comment by davkean** [comment]
@vivainio Thanks, so your expectation is that VS doesn't just automatically pick up these changes like it does with PackageReference? Do you expect to be able to open newly cloned repo in VS and have it "just work"? It seems like the expectation by @jackfoxy is that it does.

@Krzysztof-Cieslak Missing the relevance to this issue? 

@forki That is one approach albeit a bit annoying to have two paths, at what point to do you "generate" this file? If during the build, in worse case this will result in 3 design-time builds before you get full IntelliSense, which might seem to work okay for a single project, does not work well for a reasonable sized solution.

**Comment by jackfoxy** [comment]
@davkean frankly, I have no expectations. This is all black magic to me. When I find what works, I'm happy.

**Comment by vivainio** [comment]
@davkean yes, for freshly cloned repository I'm totally fine with having to do command line build (that does restore) before being able to compile any of the solutions within the repository. That's how other systems (like npm) work anyway.

**Comment by forki** [comment]
I don't think what @jackfoxy describes is completely due to paket
integration. If you restore from cmd line and open VS afterwards as
described then everything is already restored and references are properly
populated. No manual build is needed. I think that must be something else.

Ville Vainio <notifications@github.com> schrieb am Fr., 29. Juni 2018,
07:58:

**Comment by forki** [comment]
@vivainio I agree running cmd line build after clone is always a good idea. The package restore is usually only the smallest of the things that need to be prepared. Most of the projects that I have need so much more. Btw: even visualfsharp needs a command line preparation build, so that you have proto compiler ready. 

Anyway, if you keep that habit then the proposed change will probably have no downside for you. During that first restore all props files will be generated and when you start VS everything should be done already. 

@davkean regarding large solutions. I'm not completely sure if the new approach will actually be bad. That depends a bit on timing and I need to check it. This first project that is restoring in background build will restore for all projects in the solution. So all props files will

**Comment by forki** [comment]
@vivainio I agree running cmd line build after clone is always a good idea. The package restore is usually only the smallest of the things that need to be prepared. Most of the projects that I have need so much more. Btw: even visualfsharp needs a command line preparation build, so that you have proto compiler ready. 

Anyway, if you keep that habit then the proposed change will probably have no downside for you. During that first restore all props files will be generated and when you start VS everything should be done already. 

@davkean regarding large solutions. I'm not completely sure if the new approach will actually be bad. That depends a bit on timing and I need to check it. This first project that is restoring in background build will restore for all projects in the solution. So all props files will

**Comment by eiriktsarpalis** [comment]
We've been using paket in combination with the new project system and have had no significant problems, at least nothing that cannot be fixed by triggering a build. This is using a repo with 6 solutions and ~60 projects.

**Comment by forki** [comment]
@jackfoxy there is a prerelease version available. please report all bugs. (since we did not fix cache invalidation after paket's own upgrade: you need to remove paket-files folder the very first time after you upgrade)

**Comment by forki** [comment]
easiest way to test the prerelease is to put `version 5.174.0-alpha001` on first line of deps file

**Comment by jackfoxy** [comment]
by deps file you mean paket.dependencies?

**Comment by forki** [comment]
Yes

Jack Fox <notifications@github.com> schrieb am Fr., 29. Juni 2018, 16:30:

**Comment by jackfoxy** [comment]
I will likely have some sort of report by Monday.

**Comment by forki** [comment]
@jackfoxy I think you need to run paket install in order to update the targets file.

**Comment by jackfoxy** [comment]
@forki I ran paket restore and it updated the targets file. 
I just ran paket install and it did not change. So restore appears to work too.

**Comment by jackfoxy** [comment]
@forki I have not encounter any problems with the alpha release. 
I cannot determine if it had any impact on my issue. I still cannot reliably replicate the problem on demand, or replicate how to fix it. And it always does eventually get fixed.

You can close this if you are satisfied with the alpha release testing.

@cartermp  I'll probably set aside attempting a repeatable report of my problem until Preview 5,

**Comment by forki** [comment]
@jackfoxy personally I'm not convinced that paket is the reason. in most (all?) former cases when intellisense issues were reported, we usually found a way to repro without paket as well.

**Comment by jackfoxy** [comment]
@forki  I'm pretty sure you are right. I opened this issue only to make sure I had covered all possibilities. Glad the issue was useful for something I did not expect.

**Comment by jackfoxy** [comment]
I had to restore intellisense functionality in VS 15.8 Preview 3 again yesterday. I still do not have a good reproducible way to restore intellisense. It seems to be a combination of clearing obj folders, "restoring" (in a generic sense) with paket, and then opening and closing VS and attempting to force intellisense by making it do a dot completion.

This is only anecdotal, not scientific, but I think I eventually get it right by starting with

git clean -xfd
paket update -f

whereas

git clean -xfd
paket restore -f

does not work. And this is with paket alpha.

**Comment by forki** [comment]
Actually gut clean + just opening vs should already give you intellisense.

Jack Fox <notifications@github.com> schrieb am Di., 3. Juli 2018, 17:26:

**Comment by jackfoxy** [comment]
I'll try that next time.

**Comment by jackfoxy** [comment]
Amazing. This time I only did ````paket restore -f```` with the alpha.

I still had to open and close VS twice to get intellisense working. If this consistently works it's a big step forward, and I can at least report a VS bug with some consistent work-around. Thanks.

**Comment by forki** [comment]
If you did restore before opening VS, then everything works as vanilla
nuget. There is no paket in memory fiddling going on anymore. All
PackageReferences are real PackageReference tags in the paket.props file.
You can inspect that file in your /obj folder. If that file contains the
correct PackageReference tags then it's up to VS to provide intellisense
for you.

Jack Fox <notifications@github.com> schrieb am Di., 3. Juli 2018, 23:10:

**Comment by jackfoxy** [comment]
This is starting to make more sense to me now. In other words, VS does not know when paket needs to do a new restore. VS is classic nuget centric.

My most recent incident triggered when I added a new table to the database FSharp.Data.SqlProgrammabilityProvider is monitoring. The TP caches the DB schema, so when you change the schema you have to restart VS. I also added code to process the new table. Somehow this broke intellisense requiring a paket restore and 2 more VS restarts.

**Comment by forki** [comment]
> In other words, VS does not know when paket needs to do a new restore.

No, whenever a nuget restore is done, then targets files call paket to do the actual restore. It's not bound to nuget at all. Also we now generate real PackageReferences in the props file. So it IS Nuget now.

Now that you mention type providers, I think we now moving to real root cause.

**Comment by jackfoxy** [comment]
> Now that you mention type providers, I think we now moving to real root cause.

Except the other solution that exhibits this problem, FSharpx.Collections test projects, have no TPs. They do have Expecto and FsCheck, which could be bringing along some strange baggage.

**Comment by jackfoxy** [comment]
I haven't experienced this in collections with the new alpha release, yet.
